### PR TITLE
:new: Integrating database restart to replay_ethereum

### PR DESCRIPTION
**Pull request by tongzhi1998**
- Add database restart support for `replay_ethereum`
- Refactor behave tests to adapt to the new command line arguments
- Add different logger instantiation in `replay_ethereum`

### CODE REVIEW

File: src/monad/execution/ethereum/replay_ethereum.cpp [hunk 1]
```diff
@@ -73,13 +73,16 @@ int main(int argc, char *argv[])
         monad::log::logger_t::create_logger("txn_logger");
     [[maybe_unused]] auto *state_logger =
         monad::log::logger_t::create_logger("state_logger");
+    [[maybe_unused]] auto *change_set_logger =
+        monad::log::logger_t::create_logger("change_set_logger");
```
Comment (tongzhi1998) [inline src/monad/execution/ethereum/replay_ethereum.cpp:78]: Sorry @rgarc I told you yesterday that your changes shouldn't impact replay_ethereum. Turns out it isn't the case. So I fix it here. Thanks for bringing it up!

Comment (rgarc) [reply src/monad/execution/ethereum/replay_ethereum.cpp:78]: Yeah sorry about that. Just a heads up that you might have to do something similar again if this gets merged.
https://github.com/monad-crypto/monad/blob/b9ea5a237560bad485feb5e00e271e181e51765c/include/monad/execution/evmone_baseline_interpreter.hpp#L36-L38

Comment (tongzhi1998) [reply src/monad/execution/ethereum/replay_ethereum.cpp:78]: Cool!

File: include/monad/db/prepare_state.hpp [hunk 2]
```diff
@@ -69,4 +69,53 @@ prepare_state(TDB const &db, uint64_t block_number)
...
+        if (child_path.stem().string() != "CURRENT") {
+            sub_dir_cnt++;
+            latest_block_number = std::max(
+                latest_block_number, std::stoul(child_path.stem().string()));
+        }
+    }
+
+    if (sub_dir_cnt == 0u) {
+        return 0u;
+    }
```
Comment (zander-xyz) [inline include/monad/db/prepare_state.hpp]: The logic you added here to determine the block number seems better in it's own free standing function `auto_detect_starting_block_number`. Then in the `RocksTrieDB` constructor, you can directly call into the constructor specifying the block number.

Side effect here is that there's less duplicated code.

File: include/monad/db/prepare_state.hpp [hunk 3]
```diff
@@ -69,4 +69,53 @@ prepare_state(TDB const &db, uint64_t block_number)
...
+    if (fs::exists(current_dir)) {
+        fs::remove_all(current_dir);
+    }
+
+    fs::create_directories(current_dir);
+
+    auto const path = current_dir / monad::db::as_string<TDB>();
+
+    uint64_t latest_block_number = 0u;
+    uint64_t sub_dir_cnt = 0u;
```
Comment (zander-xyz) [inline include/monad/db/prepare_state.hpp]: After we break this logic out into a separate function, just return `latest_block_number` directly rather than maintaining `sub_dir_cnt`

File: include/monad/db/prepare_state.hpp [hunk 4]
```diff
@@ -69,4 +69,53 @@ prepare_state(TDB const &db, uint64_t block_number)
...
+    for (auto const &entry : std::filesystem::directory_iterator(db.root)) {
+        std::filesystem::path child_path =
+            entry.path().lexically_relative(db.root);
```
Comment (zander-xyz) [inline include/monad/db/prepare_state.hpp]: Why was the `lexically_relative` part needed? Also, this can be `auto const`

File: include/monad/db/prepare_state.hpp [hunk 5]
```diff
@@ -69,4 +69,53 @@ prepare_state(TDB const &db, uint64_t block_number)
...
+        if (child_path.stem().string() != "CURRENT") {
```
Comment (zander-xyz) [inline include/monad/db/prepare_state.hpp]: Now that we have "CURRENT" used in multiple places, I think it's probably appropriate to include it at the top of this file like

```cpp
namespace detail
{
    constexpr auto CURRENT_DATABASE = "CURRENT"sv;
}
```

File: include/monad/db/rocks_trie_db.hpp [hunk 6]
```diff
@@ -126,6 +128,55 @@ namespace detail
...
+
+                MONAD_ROCKS_ASSERT(s);
+                MONAD_ASSERT(cfds.size() == cfs.size());
+
+                return db;
+            }())
+            , accounts_trie(db, cfs[1], cfs[2])
+            , storage_trie(db, cfs[3], cfs[4])
+        {
+        }
```
Comment (zander-xyz) [inline include/monad/db/rocks_trie_db.hpp]: Instead of doing this, I would encourage you to do something like
```cpp
        RocksTrieDB(std::filesystem::path root, uint64_t block_history_size)
            : RocksTrieDB(root, auto_detect_starting_block_number(*this), block_history_size) {}
```
This reduces the amount of duplicate code that we have and need to maintain.

File: include/monad/db/rocks_trie_db.hpp [hunk 7]
```diff
@@ -165,6 +216,11 @@ namespace detail
             MONAD_DEBUG_ASSERT(s.has_value());
         }
 
+        constexpr uint64_t get_start_block_number() const
+        {
+            return start_block_number;
+        }
```
Comment (zander-xyz) [inline include/monad/db/rocks_trie_db.hpp]: We eventually want to do work to add in block number into the DBInterface. That'll be a whole separate PR that I think we should not pollute this PR with. For now, can we rename this to `block_number` which returns `start_block_number`? Also it would be nice to add a TODO noting that this needs to be fixed when we add in wider `block_number` support.

File: include/monad/db/rocks_trie_db.hpp [hunk 8]
```diff
@@ -165,6 +216,11 @@ namespace detail
             MONAD_DEBUG_ASSERT(s.has_value());
         }
 
+        constexpr uint64_t get_start_block_number() const
```
Comment (zander-xyz) [inline include/monad/db/rocks_trie_db.hpp]: `[[nodiscard]]` and `noexcept` please. Also, can you please move this function above the `DBInterface implementations` comment, since this isn't that?

File: include/monad/db/prepare_state.hpp (hunk 4 above)
Comment (tongzhi1998) [reply include/monad/db/prepare_state.hpp]: We actually don't need lexically_relative once we have .stem(). Fixed it!

File: include/monad/db/prepare_state.hpp [hunk 9]
```diff
@@ -10,6 +10,8 @@
 #include <concepts>
 #include <filesystem>
 
+#include <iostream>
+
```
Comment (zander-xyz) [inline include/monad/db/prepare_state.hpp]: why is this needed?

File: src/monad/db/test/db.cpp [hunk 10]
```diff
@@ -166,21 +168,70 @@ TEST(RocksTrieDB, block_history)
...
+
+        EXPECT_EQ(
+            db.root_hash(a),
+            0x3f9802e4f21fce3d2b07d21c8f2b60b22f7c745c455e752728030580177f8e11_bytes32);
+        EXPECT_EQ(
+            db.root_hash(),
+            0x3f7578fb3acc297f8847c7885717733b268cb52dc6b8e5a68aff31c254b6b5b3_bytes32);
+    }
+
+    auto const root_for_second_constructor = test::make_db_root(
```
Comment (zander-xyz) [inline src/monad/db/test/db.cpp]: Sorry, can you make this a separate unit test instead of appending onto the existing one?

File: src/monad/db/test/db.cpp [hunk 11]
```diff
@@ -166,21 +168,70 @@ TEST(RocksTrieDB, block_history)
...
+            .account_changes = {{a, acct}},
+            .storage_changes = {{a, {{key1, value1}, {key2, value2}}}}});
+        db.create_and_prune_block_history(block_number++);
+
+        EXPECT_EQ(
+            db.root_hash(a),
+            0x3f9802e4f21fce3d2b07d21c8f2b60b22f7c745c455e752728030580177f8e11_bytes32);
+        EXPECT_EQ(
+            db.root_hash(),
+            0x3f7578fb3acc297f8847c7885717733b268cb52dc6b8e5a68aff31c254b6b5b3_bytes32);
```
Comment (zander-xyz) [inline src/monad/db/test/db.cpp:212]: we also have `block_number()` so should we test that as well? ie. `db.block_number() == <my_exected_block_number>`

File: include/monad/db/prepare_state.hpp (hunk 9 above)
Comment (tongzhi1998) [reply include/monad/db/prepare_state.hpp]: I was using it to debug but forget to delete. Sorry about that.
